### DISCUSSION TITLE
Questions about MolToSmiles

### DISCUSSION THREAD

**Discussion by zengkaipeng** [rdkit/rdkit#8003] [Q&A] [answered]
When we use MolToSmiles with `doRandom=True`, what's the difference between setting the `canonical` parameter to True and False? Both sets of parameters seem to generate randomized SMILES. I would like to know if someone could provide some examples for me. Thanks in advance.

**Comment by kienerj** [comment]
Try 

```
for i in range(10):
    print (Chem.MolToSmiles(m, canonical=False))
```

vs

```
for i in range(10):
    print (Chem.MolToSmiles(m, doRandom=True))
```

The first gives a constant output that I assume depends on atom indexes, how the molecule was created from the input, so it is constant but not canonical.

`doRandom` gives variable output with each call, it's random, doesn't depend on the atom index / input format.

**Reply by zengkaipeng** [reply]
Thanks for your answer, but what I want is to find out the difference between
```
for i in range(10):
    print(Chem.MolToSmiles(m, canonical=True, doRandom=True))
```
and 
```
for i in range(10):
    print(Chem.MolToSmiles(m, canonical=False, doRandom=True))
```

or the two codes above have no difference.

**Reply by DavidACosgrove** [reply]
It doesn't make sense to do canonical and doRandom at the same time.  A canonical SMILES should be the same whatever the atom order when it begins - that's what it is intended to be.  This article gives a detailed description:
https://luis-vollmers.medium.com/tutorial-to-smiles-and-canonical-smiles-explained-with-examples-fbc8a46ca29f
It talks about a different canonicalising algorithm from the RDKit one, but the principle is the same.

**Reply by kienerj** [reply, marked as answer]
`doRandom` takes precedence over canonical and hence canonical is ignored. It makes sense in my opinion as you can then do `Chem.MolToSmiles(m, doRandom=True)` and get random smiles. canonical is True by default so ignoring it when doRandom=True makes sense.

**Reply by zengkaipeng** [reply]
> It doesn't make sense to do canonical and doRandom at the same time. A canonical SMILES should be the same whatever the atom order when it begins - that's what it is intended to be. This article gives a detailed description: https://luis-vollmers.medium.com/tutorial-to-smiles-and-canonical-smiles-explained-with-examples-fbc8a46ca29f It talks about a different canonicalising algorithm from the RDKit one, but the principle is the same.

I asked this question because when I was reading the MolToSmiles code, I didn't see any explicit setting that when `doRandom=True`, the `canonical` parameter is ignored. I agree that `doRandom` and `canonical` should not be set to `True` at the same time, but anyway, thank you for your answer.

**Reply by DavidACosgrove** [reply]
@zengkaipeng It's a fair point.  I have submitted a PR with clearer docstrings.
